rfc-134-govuk-wide-session-cookie-and-login.md

```diff
@@ -0,0 +1,354 @@
+
+```
+Set-Cookie: __Host-govuk_account_session=<value>; secure; httponly; samesite=strict; path=/
+```
+
+It can be expired, logging a user out, by re-setting it with
+`max-age=0`.
+
+This cookie cannot be set on the `gov.uk` domain, so service domains
+will need to re-authenticate and manage their own sessions.  This RFC
+is just about authentication and attribute use on www.gov.uk itself.
+The cross-government single-sign-on part of this work is out of scope.
+
+[domain-locked]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#Cookie_prefixes
+
+#### Cookie encryption
+
+The cookie value is encrypted with a private key.  This ensures that
+an attacker cannot tamper with the value.  Apps which consume the
+cookie can use the public key to validate it.  This key can be truly
+public, as the session ID itself is not sensitive.
+
+If the cookie value is malformed, for example if an attacker has tried
+re-encrypting it with a different private key, the consuming app
+should log the user out by expiring the cookie.
+
+#### Session duration
+
+We will have an internal API app (more on this later), keeping track
+of the mapping from session ID to user information, such as a user ID
+or OAuth tokens.
+
+This session ID will expire after 1 hour of inactivity, at which point
+the user will need to log in again.  The expiration timeout is reset
+whenever the session is used to access some personalised content.
+
+The average GOV.UK session length is 2.5 hours, but we know that this
+is distorted by a few users who leave pages open for very long times.
+We can revise the session ID timeout if we see evidence that 1 hour is
+too short.
+
+### Use custom HTTP headers to pass the cookie around
+
+[Fastly's best practices for cookies][fastly-cookies-practices]
+suggest parsing cookies into custom headers and using these headers
+for caching purposes, rather than caching based on the entire
+`Set-Cookie` header (which in our case also contains
+non-account-related things like A/B test bucket assignment, Google
+Analytics session ID, and cookie consent preferences).
+
+In addition, we have nginx configuration to strip out `Set-Cookie`
+headers.  By using custom headers we keep that logic in place,
+ensuring we can't make a mistake and start setting a cookie from one
+of our apps inappropriately.
+
+We will introduce two new headers:
+
+- `GOVUK-Account-Session`: holds the encrypted session ID.  Set by
+  Fastly to pass the cookie to our apps, set by our apps to create a
+  new cookie.
+
+- `GOVUK-Account-End-Session`: set by our apps to expire the cookie.
+
+We will need to make two changes to our VCL.
+
+[fastly-cookies-practices]: https://developer.fastly.com/reference/http-headers/Cookie/#best-practices
+
+#### Changes to `vcl_recv`
+
+When receiving a request, extract the cookie value and pass it in the
+header:
+
+```vcl
+set req.http.GOVUK-Account-Session = req.http.Cookie:__Host-govuk_account_session;
```

Comment:
What happens if the cookie isn't present? We should unset the `GOVUK-Account-Session` header so a malicious actor can't set that in a custom header, only the expected cookie should set it.
```suggestion
unset req.http.GOVUK-Account-Session;
set req.http.GOVUK-Account-Session = req.http.Cookie:__Host-govuk_account_session;
```

Reply:
I think that if the cookie isn't present the current code is equivalent to `set req.http.GOVUK-Account-Session = ""`, but we can check that